schema/ClaimsObject.md

```diff
@@ -0,0 +1,67 @@
+# Claims Object
+
+## What is this?
+
+The Claims Object is a JSON object used to ensure that TDF Clients
+attempting to decrypt an encrypted file under the Trusted Data Format
+has all permissions required to do so.
+
+## How does it work?
+
+When an TDF Client wishes to decrypt a file, the following steps using
+the Claims Object are made:
+
+1. The client requests a Claims Object by first authenticating via the
+OpenID Connect (OIDC) Identity Provider (IdP) with Custom Claims
+support (in this case Keycloak), and if authentication succeeds, a
+[TDF Claims Object](../schema/ClaimsObject.md) is obtained from
+Attribute Provider and signed by the IdP.  The signed OIDC JWT is
+returned to the client with the Claims Object inside. The Claims
+Object contains [ClaimsAttributes](ClaimsAttributeObject.md) to which
+the TDF Client has access.
+
+
+2. The client requests a decrypt from the Key Access Server (KAS), 
+presenting the OIDC JWT (containing the Claims Object) as a bearer token to the KAS. 
+The KAS ensures that the requestor has the correct permissions to access
+the contents of the file by:
+- Examining the validity of the OIDC JWT signature.
+- Determining if the TDF Client has all the required Attributes and is on
+  the [Policy Object](PolicyObject.md).
+
+If these requirements are met, the KAS will permit a decrypt of the file.
+
+
+## Example
+
+```javascript
+{
+  "userId": "user@virtru.com",
+  "aliases": [],
+  "attributes": [
+    {
+      "obj": {"attribute": "https://example.com/attr/Classification/value/S"}}
+    },
+    {
+      "obj": {"attribute": "https://example.com/attr/COI/value/PRX"}}
+    }
+  ],
+  "publicKey": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAy18Efi6+3vSELpbK58gC\nA9vJxZtoRHR604yi707h6nzTsTSNUg5mNzt/nWswWzloIWCgA7EPNpJy9lYn4h1Z\n6LhxEgf0wFcaux0/C19dC6WRPd6 ... XzNO4J38CoFz/\nwwIDAQAB\n-----END PUBLIC KEY-----",
+  "schemaVersion:": "x.y.z"
+}
+```
+
+|Parameter|Type|Description|Required?|
+|---|---|---|---|
+|`userId`|String|An id used to identify the entity, such as email.|Yes|
+|`aliases`|Array|`aliases` are not yet implemented, but will provide further flexibility by assigning additional aliases to a user. This can remain empty for the time being.|Yes|
```

Comment:
Should we remove these?

Reply:
They're still there in the code so not yet - but it's already clear that `userId` at least probably needs to go, it's largely useless in an OIDC context - we don't use it for any of the OIDC auth flows, not all SDK consumers are `users`, and I personally believe it's redundant with `clientId` in an OIDC context.

Reply:
Again, we left all of the Entity Object attributes that aren't involved in signatures intact because we haven't touched any of that downstream code.  We did not have a reason to.  If we need to, I think we can address that later after this PR.

Reply:
I'm not convinced that having an aspirational spec is a bad thing - having a major version 4.0.0 lets us break with the past.

But specific to this issue, aliases are preserved in the code in most clients, but several clients don't use them (I'm pretty sure only C++ has implemented both setters and getters), and IIRC the openstack backend ignores them entirely and I don't see any reference to them in tdf3-kas-virtru, either, so I don't think they are persisted anywhere.

Reply:
~I'll create a ticket to address removing `userId/alias` and updating the spec with the code.~

https://virtru.atlassian.net/browse/PLAT-1206